go-interface-tutorial/main.go

```diff
@@ -34,48 +37,102 @@ func (sdb *ShopDB) CountSales(since time.Time) (int, error) {
 	return count, err
 }
 
+func (sdb *ShopDB) CreateBooks() (bool, error) {
+	tableCheck, err := sdb.Query("SELECT * FROM books;")
+
+	if tableCheck == nil {
+		_, err := sdb.Query("CREATE TABLE books (title VARCHAR(50) PRIMARY KEY)")
+
+		return false, err
+	} else {
+		return true, err
+	}
+}
+
 func main() {
-	db, err := sql.Open("postgres", "postgres://user:pass@localhost/db")
+	connStr := "host=localhost port=5400 user=docker password=docker sslmode=disable"
+	db, err := sql.Open("postgres", connStr)
 	if err != nil {
 		log.Fatal(err)
 	}
 	defer db.Close()
 
 	shopDB := &ShopDB{db}
-	sr, err := calculateSalesRate(shopDB)
+	//sr, err := calculateSalesRate(shopDB)
+	sr, err := createBooks(shopDB)
 	if err != nil {
 		log.Fatal(err)
 	}
-	fmt.Printf(sr)
-}
+	fmt.Printf("%s", sr)
 
-// Swap this to use the ShopModel interface type as the parameter, instead of the
-// concrete *ShopDB type.
-func calculateSalesRate(sm ShopModel) (string, error) {
-	since := time.Now().Add(-24 * time.Hour)
-
-	sales, err := sm.CountSales(since)
-	if err != nil {
-		return "", err
-	}
+	//fmt.Println(PostgresRepository.searchTable)
+}
 
-	customers, err := sm.CountCustomers(since)
+func createBooks(sm ShopModel) (string, error) {
+	books, err := sm.CreateBooks()
 	if err != nil {
 		return "", err
 	}
 
-	rate := float64(sales) / float64(customers)
-	return fmt.Sprintf("%.2f", rate), nil
+	return fmt.Sprintf("%t", books), nil
 }
 
+// Swap this to use the ShopModel interface type as the parameter, instead of the
+// concrete *ShopDB type.
+// func calculateSalesRate(sm ShopModel) (string, error) {
+// 	since := time.Now().Add(-24 * time.Hour)
+
+// 	sales, err := sm.CountSales(since)
+// 	if err != nil {
+// 		return "", err
+// 	}
+
+// 	customers, err := sm.CountCustomers(since)
+// 	if err != nil {
+// 		return "", err
+// 	}
+
+// 	rate := float64(sales) / float64(customers)
+// 	return fmt.Sprintf("%.2f", rate), nil
+// }
+
 type Repository interface {
 	search(query string) []string
 }
 
+type PostgresRepository struct {
+	sdb *ShopDB
+}
+
 type FakeDBRepository struct {
 	searchQuery []string
 }
 
+func (p PostgresRepository) search(query string) []string {
+	rows, err := p.sdb.Query("SELECT * FROM books WHERE title='" + query + "'")
```

Comment:
We need to prevent sql injection here, so please read up on how to use placeholder parameters.